lib/bloom_remit_client/requests/url_builder.rb

```diff
@@ -0,0 +1,56 @@
+module BloomRemitClient
+  module Requests
+    class UrlBuilder
+      require 'addressable/uri'
+      include Virtus.model
+      include ActiveModel::Validations
+
+      attribute :scheme, String, default: 'https'
+      attribute :subdomain, String
+      attribute :userinfo, String
+      attribute :path, String
+      attribute :host, String
+      attribute :port, Integer
+      attribute :query, String
+      attribute :query_params, Hash, default: {}
+      attribute :fragment, String
+
+      validates :path, :host, :scheme, presence: true
+      validates :scheme, inclusion: ['https', 'http']
+
+      def call
+        builder.build(to_h).to_s
+      end
+
+      def host
+        return super unless subdomain.present?
+        "#{subdomain}.#{super}"
+      end
+
+      def path
```

Comment:
I found it much easier to test when I could set the attributes directly. Using Virtus' lazy execution allowed for that. Looks like you addressed it already though.